Added Vagrant configuration

11galpid opened the pull request:
-Added .gitignore to ignore .vagrant folder created after using Vagrant.
-Added initialization scripts (both root and usermode scripts)
-Added startup (currently empty) script
-Added Vagrantfile with updated configuration

11galpid committed:
Initial Vagrant Configuration

-Added .gitignore to ignore .vagrant folder created after using Vagrant
and desktop.ini (windows junk file).
-Added initialization scripts (both root and usermode scripts) used to
setup both priviledges and unpriviledged firsts-time scripts.
-Added pipenv file to be used for easy track of required python packages
-Added Vagrantfile which includes:
->VM Specification (Cores and RAM)
->Fedora/cloud-based
->Shared folder (2-way)
->Forwarding a port to be used later by Django

lmilbaum commented on Vagrantfile:
```diff
@@ -0,0 +1,15 @@
+  	v.cpus = 2
+  end
+  config.vm.synced_folder "init_scripts/", "/init_scripts", create: false
+  config.vm.synced_folder "selotape/", "/home/vagrant/selotape", create: true
```
The project files are synced automatically to /vagrant.  What is the purpose of syncing those files again?

11galpid replied:
From my usage, the /vagrant folder wasn't syncing properly (Uses RSYNC by default), reading online a fix is to add a different folder, and it by default uses the VirtualBox syncing method, which works great

lmilbaum replied:
Take a look at this option: https://developer.fedoraproject.org/tools/vagrant/vagrant-sshfs.html

11galpid replied:
Thanks, this looks good, but its similar to the configuration I've set before, just uses a different method for the file system of the storage (ssh-fs).
Anyway, I'm updating the machine to Ubuntu (as described in one of the other comments), and the default /vagrant folder works fine in there, so we can remove this shared folders in the next push.

lmilbaum commented on Vagrantfile:
```diff
@@ -0,0 +1,15 @@
+  end
+  config.vm.synced_folder "init_scripts/", "/init_scripts", create: false
+  config.vm.synced_folder "selotape/", "/home/vagrant/selotape", create: true
+  config.vm.provision "shell", privileged: true,
```
Is privileged necessary ?

11galpid replied:
Yes, this is meant to be run as root (system level configuration)

lmilbaum replied:
The provisioning is already executed as root as far as I know.

11galpid replied:
Yes, but because of the fact that one of the scripts needs root and the other doesn't, I feel like it should be explicitly stated.

ifireball replied:
@lioramilbaum can we resolve this conversation?

lmilbaum commented on Vagrantfile:
```diff
@@ -0,0 +1,15 @@
+  config.vm.synced_folder "init_scripts/", "/init_scripts", create: false
+  config.vm.synced_folder "selotape/", "/home/vagrant/selotape", create: true
+  config.vm.provision "shell", privileged: true,
+    inline: "/bin/sh /init_scripts/init-root.sh"
```
You might want to consider using the path attribute

lmilbaum replied:
Don't forget to add a comment or/and emoji (thumb up) in case you have decided to follow the review comment.

lmilbaum commented on init_scripts/init-root.sh:
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
```
Please elaborate what this script is doing and why it is needed

11galpid replied:
Disables built-in repositories which seemed to slow down the "dnf update" procedure, adding a faster localized repository (israeli based).
Then the script actually updates the system, and installs system-level tools which can come in handy for the developer (as this cloud-based image is stripped, even "vim" is not included)_

lmilbaum replied:
There is a discussion over Slack whether to use a Fedora box or Ubuntu. You might want to follow that.

alonweissfeld replied:
[From the Vagrant docs](https://www.vagrantup.com/docs/boxes#official-boxes), the HashiCorp (makers of vagrant) and Bento boxes are the only two recommended box sets.  

I've set up the new `bento/ubuntu-20.04` vm box and the sync folder works perfectly -  no need to re-`vagrant up` to copy new data from the synced folder, and it's already comes with essential tools such as vim. I think we should use this box (or a different long time support box) to help us simplify our configuration at the moment.

Also, this box is [supported by GitHub Actions](https://docs.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners) as discussed in slack.

11galpid replied:
From my testing, it has a few benefits, but also a few drawbacks.
Benefits:
- /vagrant works without any need for additional configuration
- Future proof for use with Github Actions

Drawbacks:
- "Heavy" box relative to the "cloud-base" (5min+- to download compared to 10seconds)
- Still has packages missing ("python3-pip" for instance is missing)

Overall, it seems worth the move for the Github Actions, so i'll push an update that uses this, but changes are gonna be made to all of the files included to support this.

ifireball replied:
@11galpid , @lioramilbaum I think we can resolve this conversation for now no?

lmilbaum commented on selotape/startup.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
```
Do not add extra code/files as place holders.

11galpid replied:
OK

alonweissfeld commented on init_scripts/init-root.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+echo "nameserver 8.8.8.8" | sudo tee /etc/resolv.conf
```
Can you elaborate why are we writing both to standard output and the file `resolve.conf`? What is the purpose of this file? maybe it's worth adding a comment

11galpid replied:
8.8.8.8 is Google's DNS server, this line basically resets the dns server to avoid issues with virtualboxe's default nat assigned dns server (which sometimes causes issues, as it did on my machine)

ifireball replied:
You really shouldn't mess with that - its one of those things that may break on other people's machines.

alonweissfeld reviewed (commented):
@11galpid Looks good! 👏  I think it's a good simplification change for now.

lioramilbaum commented:
When the PR is ready for a re-review, please pick the re-request review symbol next to the reviewers names

lioramilbaum commented:
@ifireball I have added you as a reviewer.  Take a look.

ifireball commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+.vagrant/
+desktop.ini
```
See the "no-newline at the end of file" warning above - please fix that - there should be an appropriate setting for your text editor.

ifireball commented on Vagrantfile:
```diff
@@ -0,0 +1,11 @@
+Vagrant.configure("2") do |config|
+  config.vm.box = "bento/ubuntu-20.04"
```
Using Ubuntu will limit our ability to assist you with system issues. Please consider using Fedora or CentOS.

alonweissfeld replied:
But as discussed in Slack, wouldn't it be better to choose a more "out of the box" box that GitHub actions support?
Correct me if I'm wrong, using Docker to run Fedora won't replace the usage of Vagrant?

ifireball replied:
Don't confuse GitHub actions and Vagrant. There is a an up to date Fedora Vagrant box that is officially released and supported by the Fedora project. For Vagrant you should use that.

With github actions - if you use the GitHub provided VMs - they are pre-installed with some Ubuntu version simply because that's what the GH engineers chose to provision their AZURE VMs with - but they've included the container support feature precisely to make that choice meaningless. 

As Red Hatters we work with Fedora, RHEL and CentOS all the time, we will be much better equipped to help you solve issues that arise on a Fedora environment then if those issue arise on Ubuntu or some other environment.

11galpid replied:
Ok then, Im moving back to fedora/32-cloud-base

ifireball commented on init_scripts/init-root.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+echo "nameserver 8.8.8.8" | sudo tee /etc/resolv.conf
+# Stay up-to-date
+apt-get -y update
```
This would make your deveopment environment not reproducible. You will get a different environment every time you run the Vagrant deployment. Please avoid doing that.

ifireball commented on init_scripts/init-root.sh:
```diff
@@ -0,0 +1,6 @@
+echo "nameserver 8.8.8.8" | sudo tee /etc/resolv.conf
+# Stay up-to-date
+apt-get -y update
+apt-get -y install python3-pip
```
you should try to use a fixed version so that you're guaranteed to have the same environment every time you start it.

ifireball commented on init_scripts/init-user.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+pip3 install django --user
```
please learn about Python virtual environments and use them to manage your Python dependencies. I recommend using one of the modern Python dependency management tools such as `Pipenv` or `Poetry`, but you can use `requirements.txt`.

11galpid replied:
Read about pipenv, will change the django install to user that (and future packages will be added to it as well).
Just a note, to install pipenv, we still need to use pip install,  however this time I use a specific pipenv version to stay consistent.

ifireball reviewed (changes requested):
This change includes many different and unrelated thongs in it - they should generally be sent in different PRs, but since it ends up being a very small Pr overall, we can accept that for now.

ifireball reviewed (changes requested):
"PR review response" is not a good commit message - will it be helpful for you when you see it in `git log` ?

I suggest you squash all the commits in this PR to a single will-documented commit.

With all the discussion threads, I've lost track of what needs work and what is ok for now,  so once you'll submit the new commit we'll review from scratch.

ifireball commented on init_scripts/init-user.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
```
this is a bad comment - comments should not describe *what* you do, comments should descry *why* you do it or things that could be not obvious about it.

The header section of a script should also include its name and details about its purpose, when, how and by whom it is executed.

ifireball reviewed (changes requested):
This change should probably also a README update telling other developers about the existence of Vagrant and how to use it to bring up the development environment.

11galpid committed:
Updated script files and README
-Added comments that explain the usage and flow better
-Readme now includes a basic step-by-step procedure for setting up the
development machine

ifireball reviewed (approved):
LGTM - pending re-review from @lioramilbaum

11galpid committed:
Resolved merge conflict

lmilbaum reviewed (approved):
LGTM
